presto-common/src/main/java/com/facebook/presto/common/block/VariableWidthBlockBuilder.java

```diff
@@ -244,9 +245,12 @@ public BlockBuilder writeLong(long value)
     @Override
     public BlockBuilder writeBytes(Slice source, int sourceIndex, int length)
     {
+        checkValidSliceRange(sourceIndex, length);
```

Comment:
This is probably too expensive. This is one of the most heavily used functions so make sure it doesn't impact common use cases.

Maybe TRY/CATCH will work better.

Reply:
> For every write call, Slice already does similar [checks](https://github.com/prestodb/slice/blob/master/src/main/java/com/facebook/slice/DynamicSliceOutput.java#L136). Also with this approach the code is much cleaner compared to try/catch way.

We need to balance cleanliness with performance in backend code. So I suggest a writing a benchmark test and measure the impact.

Reply:
@kaikalur added benchmark numers.

Reply:
> @kaikalur added benchmark numbers.

Hmm more than 15% overhead? I guess since a lot of other checks are happening it's fine. Make sure to run some verifier suites and the overhead is tolerable

Reply:
@kaikalur wondering why this check is expensive. https://github.com/prestodb/presto/pull/22063#discussion_r1513261003

Reply:
> @kaikalur wondering why this check is expensive. [#22063 (comment)](https://github.com/prestodb/presto/pull/22063#discussion_r1513261003)

Good question! Not immediately clear why.

@swapsmagic - also run with TRY/CATCH and see what overhead that has.

Reply:
Benchmark with Try/Catch:
------------------
Result "com.facebook.presto.operator.aggregation.BenchmarkArrayAggregation.arrayAggregation":
  28.105 ±(99.9%) 0.938 ns/op [Average]
  (min, avg, max) = (26.503, 28.105, 29.512), stdev = 1.080
  CI (99.9%): [27.167, 29.043] (assumes normal distribution)

Run complete. Total time: 00:01:41

Benchmark                                    (type)  Mode  Cnt   Score   Error  Units
BenchmarkArrayAggregation.arrayAggregation  VARCHAR  avgt   20  28.105 ± 0.938  ns/op

Process finished with exit code 0

Reply:
> ## Benchmark with Try/Catch:
> Result "com.facebook.presto.operator.aggregation.BenchmarkArrayAggregation.arrayAggregation": 28.105 ±(99.9%) 0.938 ns/op [Average] (min, avg, max) = (26.503, 28.105, 29.512), stdev = 1.080 CI (99.9%): [27.167, 29.043] (assumes normal distribution)
> 
> Run complete. Total time: 00:01:41
> 
> Benchmark (type) Mode Cnt Score Error Units BenchmarkArrayAggregation.arrayAggregation VARCHAR avgt 20 28.105 ± 0.938 ns/op
> 
> Process finished with exit code 0

Hmm - looks fishy. Try/catch is supposed to be zero cost unless the catch block is executed. These benchmark runs may not be reliable if you are doing something else on the side that might skew the result.

Reply:
Synced up with @kaikalur. Closing this.